src/ray/core_worker/task_event_buffer.h

```diff
@@ -0,0 +1,178 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <memory>  // std::shared_ptr
+#include <string>
+
+#include "absl/base/thread_annotations.h"
+#include "absl/synchronization/mutex.h"               // absl::Mutex
+#include "absl/types/optional.h"                      // std::unique_ptr
+#include "ray/common/asio/instrumented_io_context.h"  // instrumented_io_context
+#include "ray/common/asio/periodical_runner.h"        // PeriodicRunner
+#include "ray/common/id.h"                            // TaskID
+#include "ray/common/task/task_spec.h"                // TaskSpecification
+#include "ray/gcs/gcs_client/gcs_client.h"            // Gcs::GcsClient
+#include "src/ray/protobuf/gcs.pb.h"                  // rpc::TaskEventData
+
+namespace ray {
+namespace core {
+
+namespace worker {
+
+class TaskEventBuffer {
+ public:
+  virtual ~TaskEventBuffer() = default;
+
+  /// Add a task event with optional task metadata info.
+  ///
+  /// \param task_id Task ID of the task.
+  /// \param task_status Status of the task event.
+  /// \param task_info Immutable TaskInfoEntry of metadata for the task.
+  /// \param task_state_update Mutable task states updated for task execution.
+  virtual void AddTaskStatusEvent(
+      TaskID task_id,
+      rpc::TaskStatus task_status,
+      std::unique_ptr<rpc::TaskInfoEntry> task_info,
+      std::unique_ptr<rpc::TaskStateEntry> task_state_update) = 0;
+
+  /// Add a profile event.
+  ///
+  /// \param task_id Task id of the task.
+  /// \param event The profile event.
+  /// \param component_type Type of the component, e.g. worker.
+  /// \param component_id Id of the component, e.g. worker id.
+  /// \param node_ip_address Node IP address of this node.
+  virtual void AddProfileEvent(TaskID task_id,
+                               rpc::ProfileEventEntry event,
+                               const std::string &component_type,
+                               const std::string &component_id,
+                               const std::string &node_ip_address) = 0;
+
+  /// Flushing task events to GCS.
+  ///
+  /// This function will be called periodically configured by
+  /// `RAY_task_events_report_interval_ms`, and send task events stored in a buffer to
+  /// GCS. If GCS has not responded to a previous flush, it will defer the flushing to
+  /// the next interval (if not forced.)
+  ///
+  /// \param force When set to true, buffered events will be sent to GCS even if GCS has
```

Comment:
Why would we ever set force?

Reply:
this is when at disconnect, we would want to flush to GCS even if the previous request hasn't responded?